lib/IRGen/LoadableByAddress.cpp

```diff
@@ -0,0 +1,1531 @@
+        auto *RELI = dyn_cast<ReleaseValueInst>(currIns);
+        visitReleaseInst(RELI);
+        break;
+      }
+      case ValueKind::DebugValueInst: {
+        auto *DI = dyn_cast<DebugValueInst>(currIns);
+        visitDebugValueInst(DI);
+        break;
+      }
+      case ValueKind::SwitchEnumInst: {
+        auto *SEI = dyn_cast<SwitchEnumInst>(currIns);
+        visitSwitchEnumInst(SEI);
+        break;
+      }
+      default: {
+        assert(!ApplySite::isa(currIns) && "Did not expect an ApplySite");
+        assert(!dyn_cast<MethodInst>(currIns) && "Unhandled Method Inst");
+        visitInstr(currIns);
+        break;
+      }
+      }
+    }
+  }
+}
+
+static bool modifiableApply(ApplySite applySite, irgen::IRGenModule &Mod) {
+  // If the callee is objc_method then use the old ABI
+  if (applySite.getSubstCalleeType()->getRepresentation() ==
+      SILFunctionTypeRepresentation::ObjCMethod) {
+    return false;
+  }
+  return true;
+}
+
+void LargeValueVisitor::visitApply(ApplySite applySite) {
+  if (!modifiableApply(applySite, pass.Mod)) {
+    return visitInstr(applySite.getInstruction());
+  }
+  for (Operand &operand : applySite.getArgumentOperands()) {
+    SILValue currOperand = operand.get();
+    SILType silType = currOperand->getType();
+    if (std::find(pass.largeLoadableArgs.begin(), pass.largeLoadableArgs.end(),
+                  currOperand) != pass.largeLoadableArgs.end() ||
+        isLargeLoadableType(silType, pass.Mod) ||
+        std::find(pass.funcSigArgs.begin(), pass.funcSigArgs.end(),
+                  currOperand) != pass.funcSigArgs.end()) {
+      pass.applies.push_back(applySite.getInstruction());
+      return;
+    }
+  }
+}
+
+static bool isMethodInstAppliedToObjC(MethodInst *instr) {
+  for (auto *user : instr->getUses()) {
+    if (ApplySite::isa(user->getUser())) {
+      ApplySite applySite = ApplySite(user->getUser());
+      if (applySite.getSubstCalleeType()->getRepresentation() ==
+          SILFunctionTypeRepresentation::ObjCMethod) {
+        return true;
+      }
+    }
+  }
+  return false;
+}
+
+void LargeValueVisitor::visitMethodInst(MethodInst *instr) {
+  if (isMethodInstAppliedToObjC(instr)) {
+    // Do not change the method! ObjC ABI
+    return;
+  }
+  SILType currSILType = instr->getType();
+  CanType currCanType = currSILType.getSwiftRValueType();
+  SILFunctionType *currSILFunctionType =
+      dyn_cast<SILFunctionType>(currCanType.getPointer());
+  if (!currSILFunctionType) {
+    llvm_unreachable("unsupported type");
+  }
+  Lowering::GenericContextScope GenericScope(
+      instr->getModule().Types, currSILFunctionType->getGenericSignature());
+  if (containsLargeLoadable(currSILFunctionType->getParameters(), pass.Mod)) {
+    pass.methodInstsToMod.push_back(instr);
+    return;
+  }
+}
+
+void LargeValueVisitor::visitStoreInst(StoreInst *instr) {
+  SILValue src = instr->getSrc();
+  if (std::find(pass.largeLoadableArgs.begin(), pass.largeLoadableArgs.end(),
```

Comment:
It feels like this code should be factored into a common helper that is called in the visits of the various instructions kinds.